jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -1545,7 +1576,8 @@ public static void appendSchemeHostPort(StringBuffer url, String scheme, String
         }
     }
 
-    public static boolean equalsIgnoreEncodings(String uriA, String uriB)
+    // Only URIUtil is using this method
+    static boolean equalsIgnoreEncodings(String uriA, String uriB)
```

Comment:
I don't see a fix in this method for handling multi-byte encodings.
Perhaps the changes in the URI method might avoid calling this method with such encoding differences, but we should still fix this method to avoid future bugs

Reply:
The `equalsIgnoreEncoding(String, String)` is not used outside of `equalsIgnoreEncoding(URI, URI)`, so it's essentially protected by the checks already done there.

I started down the path of writing support for multi-byte UTF-8 sequences properly in `equalsIgnoreEncoding(String, String)` and quickly discovered that we would have to decode the entire input `String` into a `Utf8StringBuilder` properly, and not check character by character like we do now in this implementation.

This is because we can easily have multiple multi-byte sequences in a row.

Examples

```
// Expecting true as a result
URIUtil.equalsIgnoreEncoding("/path/to/bä€ãm/", "/path/to/b%C3%A4%E2%82%AC%C3%A3m/");

// Expecting false as a result
URIUtil.equalsIgnoreEncoding("/path/to/bãm/", "/path%2Fto/b%C3%A3m/");
URIUtil.equalsIgnoreEncoding("/path/bãm/or/bust", "/path/b%C3%A3m/or%2Fbust");
URIUtil.equalsIgnoreEncoding("/path/to/bä€ãm/", "/path%2Fto/b%C3%A4%E2%82%AC%C3%A3m/");
```

At first, I thought this implementation would literally turn into something similar to ...

``` java
    static boolean equalsIgnoreEncodings(String uriA, String uriB)
    {
        return canonicalPath(uriA).equals(canonicalPath(uriB));
    }
```

But that would strip path params and also decode `%2F` sequences, so I didn't do that.

As a result of this test, I added a testcase for path params in the false tests.

Can't use `canonicalPath(String)` either, as that uses the `isSafe(int)` checks for encoding characters (or not), I need the reverse.

To implement this properly, I would need to decode everything that is safe to decode.
Oh-kay, i'll implement a `URIUtil.decodeSafe(String)`

Reply:
Turned out to be `String URIUtil.ensureSafeEncoding(String)` instead (with test cases)